core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -73,37 +73,34 @@ public void processExpiration() {
                log.trace("Purging data container of expired entries");
                start = timeService.time();
             }
-            int offset = 0;
             // We limit it so there is only so many async expiration removals done at the same time
-            CompletableFuture[] futures = new CompletableFuture[MAX_ASYNC_EXPIRATIONS];
+            List<CompletableFuture> futures = new ArrayList<>(MAX_ASYNC_EXPIRATIONS);
             long currentTimeMillis = timeService.wallClockTime();
-            for (Iterator<InternalCacheEntry<K, V>> purgeCandidates = dataContainer.iteratorIncludingExpired();
-                 purgeCandidates.hasNext();) {
-               InternalCacheEntry<K, V> e = purgeCandidates.next();
-               if (e.canExpire()) {
+            dataContainer.forEachExpired((ice, segment) -> {
+               if (ice.canExpire()) {
                   // Have to synchronize on the entry to make sure we see the value and metadata at the same time
                   boolean expiredMortal;
                   boolean expiredTransient;
                   V value;
                   long lifespan;
                   long maxIdle;
-                  synchronized (e) {
-                     value = e.getValue();
-                     lifespan = e.getLifespan();
-                     maxIdle = e.getMaxIdle();
-                     expiredMortal = ExpiryHelper.isExpiredMortal(lifespan, e.getCreated(), currentTimeMillis);
-                     expiredTransient = ExpiryHelper.isExpiredTransient(maxIdle, e.getLastUsed(), currentTimeMillis);
+                  synchronized (ice) {
```

Comment:
I wonder if this synchronization is still necessary

Reply:
I believe so. The problem that required this is when a user is doing a put that updates an existing entry, that we could see a partial update to the metadata, causing a premature expiration iirc.

Reply:
Right, I was probably thinking about `compute()` :)

Reply:
Yeah I agree with `compute`